src/ProgressLogging.jl

```diff
@@ -103,20 +282,54 @@ Optional first argument `name` can be used to change the name of the
 progress bar.  Additional keyword arguments are passed to `@logmsg`.
 """
 macro logprogress(name, progress = nothing, args...)
+    name_expr = :($Base.@isdefined($_name_var) ? $_name_var : "")
     if progress == nothing
         # Handle: @logprogress progress
         kwargs = (:(progress = $name), args...)
-        name = _name_var
+        progress = name
+        name = name_expr
     elseif isexpr(progress, :(=)) && progress.args[1] isa Symbol
         # Handle: @logprogress progress key1=val1 ...
         kwargs = (:(progress = $name), progress, args...)
-        name = _name_var
+        progress = name
+        name = name_expr
     else
         # Otherwise, it's: @logprogress name progress key1=val1 ...
         kwargs = (:(progress = $progress), args...)
     end
+
+    id_err = "`@logprogress` must be used inside @withprogress or `_id` keyword argument"
+    id_expr = :($Base.@isdefined($_id_var) ? $_id_var : $error($id_err))
+    for x in kwargs
+        if isexpr(x, :(=)) && x.args[1] === :_id
+            id_expr = :($asuuid($(x.args[2])))
+            # last set wins; so not `break`ing
+        end
+    end
+
+    @gensym id_tmp
+    # Emitting progress log record as old/open API (i.e., using
+    # `progress` key) and also new API based on `Progress` type.
+    msgexpr = :($ProgressString($_asprogress(
+        $name,
+        $id_tmp,
+        $_parentid_var;
```

Comment:
I'm not sure I followed everything which was going on with interpolation and macro hygiene here!

Reply:
I think I had a trouble when using a macro inside a macro. I tend to give up fighting hygiene and `esc`ape the whole block when it happens. I'll see if I can avoid this after writing tests.

Reply:
If you can make it neater by all means that would be great.

But I agree that sometimes manual escaping is just easier :disappointed:

Reply:
So, I tried to avoid interpolation but it didn't work. Here is a MWE:

```julia
julia> macro f(x)
           esc(:($Base.@isdefined $x))
       end
@f (macro with 1 method)

julia> macro g(x)
           :(@isdefined $(esc(x)))
       end
@g (macro with 1 method)

julia> let a = 1
           @f a
       end
true

julia> let a = 1
           @g a
       end
ERROR: LoadError: MethodError: no method matching @isdefined(::LineNumberNode, ::Module, ::Expr)
Closest candidates are:
  @isdefined(::LineNumberNode, ::Module, ::Symbol) at reflection.jl:265
in expression starting at REPL[26]:2
```

We could use `Expr(:isdefined, var)` (or `Expr(:islocal, var)`) directly like `Base` is doing:
https://github.com/JuliaLang/julia/commit/4800158ef513a39b380ef7ce0fa36a7ed08c0701#diff-121162110854f09c3be5b39209646009L353

But I prefer sticking with the high-level API `@isdefined` since `Expr(:isdefined, :x)` is listed only at lowered AST section: https://docs.julialang.org/en/latest/devdocs/ast/#Lowered-form-1

Reply:
> But I prefer sticking with the high-level API `@isdefined`

Yes the special forms without surface syntax are implementation details of `Base` so best to avoid them.

I expect the error here is because nested macros see the `Expr(:esc, ...)` from the parent macro and `@isdefined` can't deal with that. It's an unsolved problem (https://github.com/JuliaLang/julia/issues/23221 and all that).